README.md
# ESP8266_Artnet_DMX_DC - Artnet to DMX with dimmer curves / mappings 

This project is based on Robert Oostenvelds [esp8266_artnet_dmx512](https://github.com/robertoostenveld/arduino/tree/master/esp8266_artnet_dmx512).
It is not forked, but copied, as `esp8266_artnet_dmx512` is part of the big repository [arduino](https://github.com/robertoostenveld/arduino).

ESP8266_Artnet_DMX_DC is the extended version of esp8266_artnet_dmx512 mainly to implement [dmxmapping](https://github.com/Lutzion/dmxmapping).

## Features
- receive Art-Net data packets and output them to DMX.
- use dmxmapping to change values of the dmx channels in realtime.  
This is called dimmercurve in many dmx softwares.
- receive E1.31 / sACN alternative to Art-Net
- all features working in AP-mode
- firmware update by service menu  
  
## Sequence check
In addition i added checks for artnet sequence, which is important for all sorts of animations.

In UDP communication packets might change their order.  
When sequence is not checked, movements can make unwanted jumps.  
  
![Monitor!](./media/Monitor.png "Monitor page")

If monitor shows too many sequence errors, maybe check your network configurations and dmx software settings.

Calling the web interface might create sequence errors, as the other functions stop for a short time! 

## Settings  Dimmer curves / dmx mapping and more
![Settings!](./media/Settings.png "Settings page")

- universe: universe for Art-Net / sACN. For both cases the first universe is 0.
- channels: number of dmx channels to transfer to serial output. Default is 512.
- delay: delay in milliseconds between dmx transfers. Default is 23 for 512 channels. Thus the refresh rate will be about 40 Hz.
- useMaps: global setting for activating mappings. 0: deactivated, 1 activated
- mapChan: if useMaps is 0 and mapChan is not 0 then the mapChan value is the dmx channel, that is watched to switch mapping on and off according to the value of the dmx channel. If the channels value is smaller 127, mapping is off, otherwise it is on.
- startDelay: time in seconds, that the device will wait for the connection to the router to come up.
- e131Start: if e131Start is 1 the device will also receive sACN data, else it will only receive Art-Net.
- e131Unicast: if e131Unicast is 0, sACN will receive multicast, if 1 it will receive unicast data.
- ap-psk: if ap-psk is not empty, this is the psk for ap-mode. If it is empty, the psk is the reverse part of the mac-adress in ssid.  
  
## The service menu
### Show map files
With the menu `Show map files` in the service menu you can have a look at the existing map files.  
  
![DirMap!](./media/DirMapFiles.png "Directory of map files")

### Upload map files
The menu `Upload map files` allows it to select a map file and upload it.

## LED colors
A RGB-LED shows the different states of the esp device:
- RED: not connected to WLAN, error reading configuration, short before reboot
- YELLOW: success reading configuration
- BLUE; web server called
- MAGENTA: no artnet data for more than a second
- GREEN: everything ok

# ATTENTION
**When calling the web interface, dmx output might stop for some seconds.  
Thus do not call the web interface while using the esp-dmx-device for a live show!**


# First start
On the first start you have to connect to the open WLAN access point **ARTNET_DC_&lt;PartOfMac&gt;**. The PSK for that net is the reverse &lt;PartOfMac&gt;. Now 
- enter or select your SSID and 
- enter the password / your PSK.

After rebooting, it will connect to your WLAN.


# Changing WLAN settings
For security reasons starting the wifi-manager to change network settings is only possible after pressing the reset-button 2 times. The wifi-manager will appear, if the device gets no connection.
To force the wifi-manager to come up even if the device is connected, press the reset-button at least 5 times. Using PersWifiManager there is no forcing. PersWifiManager is called by the menu.
  

# Building the hardware
See the [details](HARDWARE.md)  
  
  
# Compiling the software
See the [details](COMPILE.md)  
  
# using dmxmapping  
  
Click the picture to watch the video.
[![dmxmapping in action as video](./media/dmxmapping_pic2vid.jpg)](http://lutzion.de/media/dmxmapping_with_2_lasers.mp4)  
  
# DISCLAIMER

**THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.**



CHANGELOG.md
# CHANGELOG

18.11.2020
  - choose between WifiMangager and PersWifiManager (compiler switch PERS_WIFI) 

07.11.2020
  - sACN / E131 (ESPAsyncE131) as alternative receiver
  - statistics and webservice refresh for sACN
  - new parameters:
    - startDelay
    - e131Start
    - e131Unicast
    - psk

01.11.2020
  - new function __DATE__2Date()

18.08.2020
  - MultiResetDetector with validResetCallback function to show reset on LED

17.08.2020
  - show resetCounter in firmware version 

15.08.2020
  - call ArtnetPolled() after random time (max 1 seconds)
  - count resets with MultiResetDetector
  - control WifiManager with ResetButton
    - 2 times reset allows WifiManager
    - 5 times reset forces WifiManager

26.07.2020
  - gui with dark design and table in monitoring

25.07.2020
  - monitoring with webSockets

19.07.2020
  - waiting for connection to router depending on startDelay
  - SSID with ChipID (reverse as psk)

15.07.2020
  - finally LittleFS with  
     `board_build.filesystem = littlefs`  
    from version 2.6.0  
  - transferred ArtnetWifi in lib, to have  
     `protected: WiFiUDP Udp;`  
    in code  

12.06.2020
  - in ArtNetPolled() changes, for IPAdress.v4() type conversion

05.06.2020
  - prompt before deleting

04.06.2020
  - possibility to delete map-files
  - changing style sheet

29.05.2020
  - added ArtNetPolled()  
    using class `typedef struct __attribute((__packed__)) _artnetPollReply`  
  - ATTENTION:
    for easy reply possibility -> subclassed  
      class ArtnetWifiDC : public ArtnetWifi  
    that makes it neccessary to access Udp  
    so change  
     private: WiFiUDP Udp; 
    to  
     protected: WiFiUDP Udp;  
    in ArtnetWifi-library. You may have to repeat after each ArtnetWifi update  
  
    Artnet-Poll-Reply needed by programs like AuroraDMX  
     https://github.com/dfredell/AuroraDMX

26.05.2020
  - RGB LED  
    Colors:  
    - RED: no WLAN, no configuration, before reboot  
    - YELLOW: configuration read successful  
    - GREEN: normal mode  
    - BLUE: website called  
    - MAGENTA: no more Art-Net data for more than a second  
  - new file <esp8266_artnet_dmx_dc.h> for "export" of LED-functions  
  
15.05.2020
  - better feedback before / after upload, update, restart 

10.05.2020
  - new variables useMaps and mapChan for dmxmapping
  - output seqErr in %
  - dmxmapping implemented incl. upload of map-files

09.05.2020
  - a bit of gui - changing html pages
  - implement config.json in file system, to start with reasonable default values
  - in function onDmxPacket: checking universe and sequence before processing data

08.05.2020
  - new format for uptime
  

src/webserverAndOta.cpp
/*
 Original by
 https://github.com/robertoostenveld/arduino/tree/master/esp8266_artnet_dmx512
 Changes by
 https://github.com/Lutzion/
*/

#include "webserverAndOta.h"
#include "esp8266_artnet_dmx_dc.h"

#include <list>

extern ESP8266WebServer server;
extern Config config;
extern unsigned long packetCounter;

File fsUploadFile;              // a File object to temporarily store the received file

std::list<file> lstFile ;

/***************************************************************************/

static String getContentType(const String& path) {
  if (path.endsWith(".html"))       return "text/html";
  else if (path.endsWith(".htm"))   return "text/html";
  else if (path.endsWith(".css"))   return "text/css";
  else if (path.endsWith(".txt"))   return "text/plain";
  else if (path.endsWith(".js"))    return "application/javascript";
  else if (path.endsWith(".png"))   return "image/png";
  else if (path.endsWith(".gif"))   return "image/gif";
  else if (path.endsWith(".jpg"))   return "image/jpeg";
  else if (path.endsWith(".jpeg"))  return "image/jpeg";
  else if (path.endsWith(".ico"))   return "image/x-icon";
  else if (path.endsWith(".svg"))   return "image/svg+xml";
  else if (path.endsWith(".xml"))   return "text/xml";
  else if (path.endsWith(".pdf"))   return "application/pdf";
  else if (path.endsWith(".zip"))   return "application/zip";
  else if (path.endsWith(".gz"))    return "application/x-gzip";
  else if (path.endsWith(".json"))  return "application/json";
  else if (path.endsWith(".map"))   return "text/plain";  // for dmx-mapping
  return "application/octet-stream";
}

/***************************************************************************/

// configuration functions
bool initialConfig()
{
  config.universe = 0;
  config.channels = 512;
  config.delay = 23;
  config.useMaps = 0 ;
  config.mapChan = 0 ;

  config.startDelay = 60 ;
  config.e131Start = 1;
  config.e131Unicast = 0 ;
  config.psk[0] = 0 ;  // empty string

  return true;
}

bool loadConfig()
{
#ifdef VERBOSE
  Serial.println("loadConfig");
#endif

#ifdef LITTLEFS
  File configFile = LittleFS.open("/config.json", "r");
#else
  File configFile = SPIFFS.open("/config.json", "r");
#endif
  if (!configFile)
  {
#ifdef VERBOSE
    Serial.println("Failed to open config file");
#endif
    return false;
  }

  size_t size = configFile.size();
  if (size > 1024)
  {
#ifdef VERBOSE
    Serial.println("Config file size is too large");
#endif
    return false;
  }

  std::unique_ptr<char[]> buf(new char[size]);
  configFile.readBytes(buf.get(), size);
  configFile.close();

  StaticJsonDocument<300> jsonBuffer;
  DynamicJsonDocument root(1024);
  DeserializationError error = deserializeJson(root, buf.get());
  if (error){
#ifdef VERBOSE
    Serial.println("Failed to parse config file");
#endif
    return false;
  }

  JSON_TO_CONFIG(universe, "universe");
  JSON_TO_CONFIG(channels, "channels");
  JSON_TO_CONFIG(delay, "delay");
  JSON_TO_CONFIG(useMaps, "useMaps");
  JSON_TO_CONFIG(mapChan, "mapChan");
  JSON_TO_CONFIG(startDelay, "startDelay");
  JSON_TO_CONFIG(e131Start, "e131Start");
  JSON_TO_CONFIG(e131Unicast, "e131Unicast");
  S_JSON_TO_CONFIG(psk, "psk");

  return true;
}

bool saveConfig()
{
#ifdef VERBOSE
  Serial.println("saveConfig");
#endif
  DynamicJsonDocument root(300);

  CONFIG_TO_JSON(universe, "universe");
  CONFIG_TO_JSON(channels, "channels");
  CONFIG_TO_JSON(delay, "delay");
  CONFIG_TO_JSON(useMaps, "useMaps");
  CONFIG_TO_JSON(mapChan, "mapChan");
  CONFIG_TO_JSON(startDelay, "startDelay");
  CONFIG_TO_JSON(e131Start, "e131Start");
  CONFIG_TO_JSON(e131Unicast, "e131Unicast");
  S_CONFIG_TO_JSON(psk, "psk");

#ifdef LITTLEFS
  File configFile = LittleFS.open("/config.json", "w");
#else
  File configFile = SPIFFS.open("/config.json", "w");
#endif
  if (!configFile)
  {
#ifdef VERBOSE
    Serial.println("Failed to open config file for writing");
#endif
    return false;
  }
  else {
#ifdef VERBOSE
    Serial.println("Writing to config file");
#endif
    serializeJson(root, configFile);
    configFile.close();
    return true;
  }
}

/***************************************************************************/

// === finalizeWebServer()
// do the outstanding web server tasks
void finalizeWebServer(uint16_t iDelay)
{
  LedBlue() ;

  for(uint16_t i = 0; i < (iDelay / 100); i++)
  {
    server.handleClient();
    delay(100);
  }

  LedRed() ;
}

// Functions for (map)file uploads
void handleMapUploadFn()
{
  server.sendHeader("Connection", "close");
  server.sendHeader("Access-Control-Allow-Origin", "*");
  //server.send(200, "text/plain", (Update.hasError()) ? "FAIL" : "OK");
  server.sendHeader("Location", (Update.hasError()) ? "/failure.html" : "/success.html");
  server.send(303);
}

void handleMapUploadUpload(){ // upload a new file to the SPIFFS
  HTTPUpload& upload = server.upload();
  if(upload.status == UPLOAD_FILE_START)
  {
    LedBlue() ;

    String filename = upload.filename;
    if(!filename.startsWith("/")) filename = "/"+filename;
    Serial.print("handleFileUpload Name: "); Serial.println(filename);

#ifdef LITTLEFS
    fsUploadFile = LittleFS.open(filename, "w");
#else
    fsUploadFile = SPIFFS.open(filename, "w");            // Open the file for writing in SPIFFS (create if it doesn't exist)
#endif

    filename = String();
  } else if(upload.status == UPLOAD_FILE_WRITE){
    if(fsUploadFile)
      fsUploadFile.write(upload.buf, upload.currentSize); // Write the received bytes to the file
  } else if(upload.status == UPLOAD_FILE_END){
    if(fsUploadFile) {                                    // If the file was successfully created
      fsUploadFile.close();                               // Close the file again
      Serial.print("handleFileUpload Size: "); 
      Serial.println(upload.totalSize);
      server.sendHeader("Location","/success.html");      // Redirect the client to the success page
      server.send(303);
    } else {
      //server.send(500, "text/plain", "500: couldn't create file");
      server.sendHeader("Location","/failure.html");      // Redirect the client to the failure page
      server.send(303);
    }
  }
}

// Functions for firmware updates
void handleFwUpdateFn()
{
  server.sendHeader("Connection", "close");
  server.sendHeader("Access-Control-Allow-Origin", "*");
  //server.send(200, "text/plain", (Update.hasError()) ? "FAIL" : "OK");
  // Redirect the client to the failure page
  server.sendHeader("Location", (Update.hasError()) ? "/failure.html" : "/success.html");
  server.send(303);

  finalizeWebServer(2000) ;
  ESP.restart();
}

void handleFwUpdateUpload() {
  HTTPUpload& upload = server.upload();
  if (upload.status == UPLOAD_FILE_START)
  {
    LedBlue() ;

    Serial.setDebugOutput(true);
    WiFiUDP::stopAll();
    Serial.printf("Update: %s\n", upload.filename.c_str());
    uint32_t maxSketchSpace = (ESP.getFreeSketchSpace() - 0x1000) & 0xFFFFF000;
    if (!Update.begin(maxSketchSpace)) { //start with max available size
      Update.printError(Serial);
    }
  } else if (upload.status == UPLOAD_FILE_WRITE) {
    if (Update.write(upload.buf, upload.currentSize) != upload.currentSize) {
      Update.printError(Serial);
    }
  } else if (upload.status == UPLOAD_FILE_END) {
    if (Update.end(true)) { //true to set the size to the current progress
      Serial.printf("Update Success: %u\nRebooting...\n", upload.totalSize);
    } else {
      Update.printError(Serial);
    }
    Serial.setDebugOutput(false);
  }
  yield();
}

// compareFiles()
// sort files by name
bool compareFiles(file& first, file& second)
{
  String f = first.name ;
  String s = second.name ;
  f.toLowerCase() ;
  s.toLowerCase() ;

  return ( f.compareTo(s) < 0 ) ;
}

// handleDirList()
// if mapOnly = true -> show only files related to dmxmapping
// files to list, sort list, print list, clear list
void handleDirList(bool mapOnly, String argDelete)
{
#ifdef VERBOSE
  Serial.println("handleDirList");
#endif

  if (argDelete.length() > 0)
  {
    //Serial.println("handleDirList -> Loesche <" + argDelete + ">");

#ifdef LITTLEFS
    if (LittleFS.exists(argDelete))
    {
      LittleFS.remove(argDelete);
    }
#else
    if (SPIFFS.exists(argDelete))
    {
      //Serial.println("remove <" + argDelete + ">");
      SPIFFS.remove(argDelete);
    }
#endif
  }

  lstFile = std::list<file>();

  String str = "<!doctype html>"
               "<html>"
               "<head>"
                "<title>ESP Artnet DMX - Directory</title>"
                "<meta charset='utf-8'>"
                "<link href='style.css' rel='stylesheet' type='text/css'>"
                "</head>"
                "<body>"
                "<div class='content'>"
                "<h1>Directory <a href='/'>&lt;&lt;&lt;</a></h1>"

                "<table align='center'>"
                "<tr><th>Name</th><th>[bytes]</th></tr>\n";

#ifdef LITTLEFS
  Dir dir = LittleFS.openDir("/");
#else
  Dir dir = SPIFFS.openDir("/");
#endif

  while (dir.next())
  {
    String sfile = dir.fileName() ;
    sfile.toUpperCase() ;
    if (!mapOnly ||
        sfile.indexOf("MAP") != -1)
    {
      // not mapOnly -> show all, 
      // mapOnly and MAP in filename -> show dmxmapping related files
      file f = file() ;
      f.name = dir.fileName() ;
      f.size = dir.fileSize();
      //f.time = dir.fileTime();
      //f.creationTime = dir.fileCreationTime();
      lstFile.push_back(f) ;
      /*
      str = str + "<tr><td align='left'><a href=" + dir.fileName() + ">" + dir.fileName() + 
            "</a></td><td align='right'>" + dir.fileSize() +  "</td></tr>" ;
      */
    }
  }

  lstFile.sort(compareFiles);

  for (std::list<file>::iterator it=lstFile.begin() ; it != lstFile.end(); ++it)
  {
    str = str + "<tr><td align='left'><a href=" + (*it).name + ">" + (*it).name + 
          "</a></td><td align='right'>" + (*it).size + 
          (mapOnly ? "&nbsp;&nbsp;<a class='del' href='?delete=" + (*it).name + "' onclick='return confirm(\"Do you want to delete &lt;" + (*it).name + "&gt; ?\");'> X</a>" : "") +
          "</td></tr>\n" ;
  }

  lstFile.clear() ;

  str = str + "</table>"
              "</DIV>"
              "</body>"
              "</html>" ;
  server.send(200, "text/html", str);
}

// webserver functions
void handleNotFound()
{
#ifdef VERBOSE
  Serial.println(String("handleNotFound(") + server.uri() + ")");
#endif

#ifdef LITTLEFS
  if (LittleFS.exists(server.uri().substring(1)))
#else
  if (SPIFFS.exists(server.uri()))
#endif
  {
    // not defined -> use static file
    handleStaticFile(server.uri());
  }
  else
  {
    String message = "File Not Found\n\n";
    message += "URI: ";
    message += server.uri();
    message += "\nMethod: ";
    message += (server.method() == HTTP_GET) ? "GET" : "POST";
    message += "\nArguments: ";
    message += server.args();
    message += "\n";
    for (uint8_t i = 0; i < server.args(); i++)
    {
      message += " " + server.argName(i) + ": " + server.arg(i) + "\n";
    }
    // server.send(404, "text/plain", message);
    // LH -> no errors -> goto start page instead
    handleRedirect("/index"); 
  }
}

void handleRedirect(String filename)
{
  char buf[64];
  filename.toCharArray(buf, sizeof(buf));
  handleRedirect(filename);
}

void handleRedirect(const char * filename)
{
#ifdef VERBOSE
  Serial.print("handleRedirect(");
  Serial.print(filename);
  Serial.println(")");
#endif
  server.sendHeader("Location", String(filename), true);
  server.send(302, "text/plain", "");
}

bool handleStaticFile(String path)
{
#ifdef VERBOSE
  Serial.println(String("handleStaticFile(") + path + ")");
#endif

#ifdef LITTLEFS
  path = path.substring(1);
#endif

  String contentType = getContentType(path);            // Get the MIME type

#ifdef LITTLEFS
  if (LittleFS.exists(path)) {                            // If the file exists
    File file = LittleFS.open(path, "r");                 // Open it
#else
  if (SPIFFS.exists(path)) {                            // If the file exists
    File file = SPIFFS.open(path, "r");                 // Open it
#endif
    //size_t sent = 
    server.streamFile(file, contentType); // And send it to the client
    file.close();                                       // Then close the file again
    return true;
  }
#ifdef VERBOSE
  Serial.println("\tFile <" + path + "> not Found");
#endif
  return false;                                         // If the file doesn't exist, return false
}

bool handleStaticFile(const char * path) {
  return handleStaticFile((String)path);
}

void handleJSON() {
#ifdef VERBOSE
  Serial.println("handleJSON");
#endif
  String message = "HTTP Request\n\n";
  message += "URI: ";
  message += server.uri();
  message += "\nMethod: ";
  message += (server.method() == HTTP_GET) ? "GET" : "POST";
  message += "\nArguments: ";
  message += server.args();
  message += "\n";
  for (uint8_t i = 0; i < server.args(); i++) {
    message += " " + server.argName(i) + ": " + server.arg(i) + "\n";
  }
#ifdef VERBOSE
  Serial.println(message);
#endif

  // this gets called in response to either a PUT or a POST
  if (server.args() == 1 && server.hasArg("plain"))
  {
    // parse it as JSON object
    DynamicJsonDocument root(300);
    DeserializationError error = deserializeJson(root, server.arg("plain"));
    if (error) {
      handleStaticFile("/failure.html");
      return;
    }
    JSON_TO_CONFIG(universe, "universe");
    JSON_TO_CONFIG(channels, "channels");
    JSON_TO_CONFIG(delay, "delay");
    JSON_TO_CONFIG(useMaps, "useMaps");
    JSON_TO_CONFIG(mapChan, "mapChan");
    JSON_TO_CONFIG(startDelay, "startDelay");
    JSON_TO_CONFIG(e131Start, "e131Start");
    JSON_TO_CONFIG(e131Unicast, "e131Unicast");
    S_JSON_TO_CONFIG(psk, "psk");
    handleStaticFile("/success.html");
  }
  else {
    // parse it as key1=val1&key2=val2&key3=val3
    KEYVAL_TO_CONFIG(universe, "universe");
    KEYVAL_TO_CONFIG(channels, "channels");
    KEYVAL_TO_CONFIG(delay, "delay");
    KEYVAL_TO_CONFIG(useMaps, "useMaps");
    KEYVAL_TO_CONFIG(mapChan, "mapChan");
    KEYVAL_TO_CONFIG(startDelay, "startDelay");
    KEYVAL_TO_CONFIG(e131Start, "e131Start");
    KEYVAL_TO_CONFIG(e131Unicast, "e131Unicast");
    S_KEYVAL_TO_CONFIG(psk, "psk");
    handleStaticFile("/success.html");
  }
  saveConfig();
}
